google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/sdk/ManagedCloudSdkUpdater.java

```diff
@@ -0,0 +1,150 @@
+  // 20 seconds to show notification and then proceed with update.
+  private static final int UPDATE_NOTIFICATION_DELAY = 1000 * 20;
+
+  @VisibleForTesting
+  static final String SDK_UPDATER_THREAD_NAME = "managed-google-cloud-sdk-updates";
+
+  static ManagedCloudSdkUpdater getInstance() {
+    if (instance == null) {
+      instance = new ManagedCloudSdkUpdater();
+    }
+    return instance;
+  }
+
+  @VisibleForTesting
+  static void setInstance(ManagedCloudSdkUpdater instance) {
+    ManagedCloudSdkUpdater.instance = instance;
+  }
+
+  private Timer updateTimer;
+  private TimerTask sdkUpdateTask;
+
+  void activate() {
+    if (!CloudSdkServiceUserSettings.getInstance().getEnableAutomaticUpdates()) {
+      return;
+    }
+
+    if (updateTimer == null) {
+      updateTimer = new Timer(SDK_UPDATER_THREAD_NAME);
+    }
+    // cancel tasks from previous activation.
+    if (sdkUpdateTask != null) {
+      sdkUpdateTask.cancel();
+    }
+
+    sdkUpdateTask =
+        new TimerTask() {
+          @Override
+          public void run() {
+            doUpdate();
+          }
+        };
+    long lastTimeOfUpdate =
+        CloudSdkServiceUserSettings.getInstance().getLastAutomaticUpdateTimestamp();
+    long delayBeforeCheckMillis = SDK_UPDATE_INTERVAL;
+    long now = getClock().millis();
+    if (lastTimeOfUpdate != 0 && lastTimeOfUpdate < now) {
+      delayBeforeCheckMillis =
+          Math.min(
+              SDK_UPDATE_INTERVAL,
+              Math.max(0, SDK_UPDATE_INTERVAL - (getClock().millis() - lastTimeOfUpdate)));
+    }
+
+    schedule(sdkUpdateTask, delayBeforeCheckMillis, SDK_UPDATE_INTERVAL);
+  }
+
+  /** Called when managed SDK update operation (update or install) successfully completes. */
+  void notifySdkUpdate() {
+    CloudSdkServiceUserSettings.getInstance().setLastAutomaticUpdateTimestamp(getClock().millis());
+  }
+
+  @VisibleForTesting
+  Clock getClock() {
+    return Clock.systemDefaultZone();
+  }
+
+  @VisibleForTesting
+  void schedule(TimerTask timerTask, long delay, long period) {
+    updateTimer.scheduleAtFixedRate(timerTask, delay, period);
+  }
+
+  @VisibleForTesting
+  javax.swing.Timer createUiTimer(int delayMillis) {
+    return new javax.swing.Timer(delayMillis, null);
+  }
+
+  private void doUpdate() {
+    CloudSdkService cloudSdkService = CloudSdkService.getInstance();
+    if (cloudSdkService instanceof ManagedCloudSdkService) {
+      // do not show notifications and start update process if SDK is still up-to-date.
+      boolean updateFeatureEnabled =
+          ServiceManager.getService(PluginInfoService.class)
+              .shouldEnable(GctFeature.MANAGED_SDK_UPDATE);
+
+      if (updateFeatureEnabled && !((ManagedCloudSdkService) cloudSdkService).isUpToDate()) {
+        // schedule UI timer to let a user decide about proceeding with the update.
+        javax.swing.Timer uiTimer = createUiTimer(UPDATE_NOTIFICATION_DELAY);
```

Comment:
don't these notifications disappear automatically after X amount of time? Do we really need to manually control the duration its shown?

Reply:
I looked briefly and I believe I didn't find such use cases - the balloon will collapse into event log message, but the notification itself stays active (not expired). Our idea here is to wait for a short time before an update to give a user some time to consider - I think I haven't see any event of notification getting automatically stale - plus we need to disable actions after this period. I'll double check this though.

Reply:
so you mean the notification message stays up indefinitely? I was pretty sure it was the opposite: it automatically hides after a period of time. Could be wrong though.

Reply:
Balloon will definitely collapse, but notification will be in the log with active actions I believe. Checking this.

Reply:
oh. so your goal here is to expire the actions in the log then? I misunderstood.

what happens after the timer sets off?

Reply:
Notification expires (which includes disabling actions since update already starts) and update is run.